### PR TITLE
Refactored ContainerStorageManager for readability and manageability

### PR DESCRIPTION
**Pull request by prateekm**
Issues: ContainerStorageManager has gotten very unwieldy with 1000+ LOC of highly complex and invocation-order dependent logic. It also does too much by trying to manage side inputs lifecycle along with regular state stores'.
 
Changes: 
    1. Separated side inputs (and standby) related store creation and restore logic into a separate SideInputsManager class.
    2. Separated helper methods into a ContainerStorageManagerUtil class, and made them all static to ensure that they don't mutate class fields. 

Most of the PR is simply moving code from one place to another. However a couple of places that introduce semantic changes or otherwise require careful review are called inline in "Refactor Note" comments.

There is a lot more room for improvement (e.g. more javadocs, more tests, simplifying method signatures, etc). But this change should make it easier to reason about the lifecycle of side inputs for now.

Tests: Existing ContainerStorageManager unit tests pass.

### CODE REVIEW

**Comment by prateekm** [comment]
cc @bharathkk to review.

File: samza-core/src/main/scala/org/apache/samza/storage/SideInputsManager.java [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+      LOG.info("SideInput Restore complete");
+    }
+  }
+
+  public Map<TaskName, Map<String, StorageEngine>> getSideInputStores() {
+    return ImmutableMap.copyOf(this.sideInputStores);
+  }
+
+  public void shutdown() {
+    // stop all side input consumers and stores
```
Comment (mynameborat) [inline samza-core/src/main/scala/org/apache/samza/storage/SideInputsManager.java:312]: We need to set the `shouldShutdown = true` here for behavior parity across the refactor?

File: samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java [hunk 2]
```diff
@@ -273,491 +200,52 @@ public ContainerStorageManager(
...
-      LOG.error(e.getMessage());
-    }
-    if (this.hasSideInputs) {
-      startSideInputs();
+      LOG.error("Error shutting down restore executor", e);
     }
+
+    // create and restore side input stores
+    this.sideInputsManager = new SideInputsManager(
+        sideInputSystemStreams, systemFactories, activeTaskChangelogSystemStreams, storageEngineFactories, storeDirectoryPaths, containerModel, jobContext, containerContext, samzaContainerMetrics, taskInstanceMetrics, taskInstanceCollectors, streamMetadataCache, systemAdmins, serdeManager,
```
Comment (mynameborat) [inline samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java:222]: shouldn't `activeTaskChangelogSystemStreams` be replaced by `changelogSystemStreams` passed in the constructor of `ContainerStorageManager`? 

e.g., 
```
  */
  @VisibleForTesting
  Map<TaskName, Map<String, Set<SystemStreamPartition>>> getTaskSideInputSSPs(ContainerModel containerModel,
      Map<String, Set<SystemStream>> sideInputSystemStreams, Map<String, SystemStream> changelogSystemStreams) {
```
method used in CSM prior to refactor which has the caller arguments
```this.taskSideInputStoreSSPs = getTaskSideInputSSPs(containerModel, sideInputSystemStreams, changelogSystemStreams);```

**Review by mynameborat — COMMENTED** [review]
Mostly looks good to me except the two comments.

Generally, I noticed you expanded the args list for the utility methods to include instance variables which were implicitly used to be explicit. However, is there any reason not to preserve the ordering of parameters from the existing methods? Would have been much easier to read, compare and review.

File: samza-core/src/main/scala/org/apache/samza/storage/SideInputsManager.java (hunk 1 above)
Comment (prateekm) [reply samza-core/src/main/scala/org/apache/samza/storage/SideInputsManager.java:318]: Fixed, thanks!

File: samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java (hunk 2 above)
Comment (prateekm) [reply samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java]: Good catch! Fixed and verified and renamed other usages to clearly distinguish between the two.

**Review by mynameborat — APPROVED** [review]

**Pull request merged**